From 39b56ecd0c234fc06a51dac4e692706912a99f9f Mon Sep 17 00:00:00 2001
From: Andrey Ulanov <andreyu@google.com>
Date: Tue, 17 Jan 2017 17:47:18 -0800
Subject: [PATCH] Set init processes as non-dumpable

This sets the init processes that join and setup the container's
namespaces as non-dumpable before they setns to the container's pid (or
any other ) namespace.

This settings is automatically reset to the default after the Exec in
the container so that it does not change functionality for the
applications that are running inside, just our init processes.

This prevents parent processes, the pid 1 of the container, to ptrace
the init process before it drops caps and other sets LSMs.

Signed-off-by: Andrey Ulanov <andreyu@google.com>
---
 libcontainer/nsenter/nsexec.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/libcontainer/nsenter/nsexec.c b/libcontainer/nsenter/nsexec.c
index 8f37d6c..ab4d504 100644
--- a/libcontainer/nsenter/nsexec.c
+++ b/libcontainer/nsenter/nsexec.c
@@ -364,6 +364,11 @@ void nsexec(void)
 		return;
 	}
 
+	if (prctl(PR_SET_DUMPABLE, 0, 0, 0, 0) != 0) {
+		pr_perror("failed to set process as non-dumpable");
+		exit(1);
+	}
+
 	// Retrieve the netlink header
 	struct nlmsghdr nl_msg_hdr;
 	int		len;
-- 
2.11.0.483.g087da7b7c-goog

